lti_placements/canvas_placements.py

```diff
@@ -0,0 +1,261 @@
+        self.canvas = canvasapi.Canvas(canvas_url, canvas_token)
+        self.zoom_placements = ZoomPlacements(self.canvas)
+
+    def generate_lti_course_report(self,
+                                   canvas_account_id: int,
+                                   enrollment_term_ids: Union[Sequence[int], None],
+                                   add_course_ids: Union[List[int], None] = None,
+                                   published: bool = True):
+    
+        account = self.canvas.get_account(canvas_account_id)
+        # Canvas has a limit of 100 per page on this API
+        per_page = 100
+
+        # Get all published courses from the defined enrollment terms
+        courses: List[canvasapi.course.Course] = []
+        if enrollment_term_ids is not None:
+            for enrollment_term_id in enrollment_term_ids:
+                logger.info(f'Fetching published course data for term {enrollment_term_id}')
+                courses_list = list(
+                    account.get_courses(
+                        enrollment_term_id=enrollment_term_id,
+                        published=published,
+                        per_page=per_page
+                    )
+                )
+                courses += courses_list
+
+        for course in courses:
+            if add_course_ids and course.id in add_course_ids:
+                add_course_ids.remove(course.id)
+            # TODO: In the future get the total count from the Paginated object
+            # Needs API support https://github.com/ucfopen/canvasapi/issues/114
+            self.get_lti_tab(course)
+
+        # If there are course_ids passed in, also process those
+        if add_course_ids:
+            for course_id in add_course_ids:
+                self.get_lti_tab(self.canvas.get_course(course_id))
+        return None
+
+    def get_lti_tab(self, course: canvasapi.course.Course) -> None:
+        # This is a new course we're looking through
+        self.course_count += 1
+        logger.info(f"Fetching course #{self.course_count} for {course}")
+        # Get tabs and look for defined tool(s) that aren't hidden
+        tabs = course.get_tabs()
+        for tab in tabs:
+            # Hidden only included if true
+            if (tab.label.upper() in SUPPORTED_PLACEMENTS and not hasattr(tab, "hidden")):
+                self.placement_count += 1
+                # This this course as a supported course
+                self.lti_placements.append({'id': self.placement_count,
+                                            'course_id': course.id,
+                                            'account_id': course.account_id,
+                                            'course_name': course.name,
+                                            'placement_type_id': SUPPORTED_PLACEMENTS.get(tab.label.upper())
+                                            })
+
+                # TODO: Find a better way of running this just for zoom
```

Comment:
Yeah it could either take pass it as `canvasapi.tab.Tab`, which could be looped in one level higher. Seems okay!

I think the reason is I call "get_lti_tab" twice. Both from the loop for courses and the loop for additional courses. I could refactor this too and just incorporate #174.

Reply:
I'm looking at this and it's better to separate it till #174. I think I'll just rename this to "get_lti_tabs" to better reflect what it's doing.

The problem is that the call to account.get_courses actually returns canvasapi.course.Course objects. So we'll also need to do this for each additional course to turn it into a course object before looping into here. I feel like it's safer to have that separated

Reply:
Sure, I'll leave it up to you, just thought I'd share what I was thinking.